netebpfext/net_ebpf_ext_bind.c

```diff
@@ -0,0 +1,249 @@
+    for (wchar_t* position = (wchar_t*)ctx->app_id_start; position < (wchar_t*)ctx->app_id_end; position++) {
+        if (*position == '\\') {
+            last_separator = position;
+        }
+    }
+    if (*last_separator == '\\') {
+        last_separator++;
+    }
+    ctx->app_id_start = (uint8_t*)last_separator;
+}
+
+void
+net_ebpf_ext_resource_allocation_classify(
+    _In_ const FWPS_INCOMING_VALUES* incoming_fixed_values,
+    _In_ const FWPS_INCOMING_METADATA_VALUES* incoming_metadata_values,
+    _Inout_opt_ void* layer_data,
+    _In_opt_ const void* classify_context,
+    _In_ const FWPS_FILTER* filter,
+    uint64_t flow_context,
+    _Inout_ FWPS_CLASSIFY_OUT* classify_output)
+{
+    SOCKADDR_IN addr = {AF_INET};
+    uint32_t result;
+    bind_md_t ctx;
+
+    UNREFERENCED_PARAMETER(layer_data);
+    UNREFERENCED_PARAMETER(classify_context);
+    UNREFERENCED_PARAMETER(filter);
+    UNREFERENCED_PARAMETER(flow_context);
+
+    if (!ebpf_ext_attach_enter_rundown(_ebpf_bind_hook_provider_registration)) {
+        classify_output->actionType = FWP_ACTION_PERMIT;
+        goto Exit;
+    }
+
+    addr.sin_port =
+        incoming_fixed_values->incomingValue[FWPS_FIELD_ALE_RESOURCE_ASSIGNMENT_V4_IP_LOCAL_PORT].value.uint16;
+    addr.sin_addr.S_un.S_addr =
+        incoming_fixed_values->incomingValue[FWPS_FIELD_ALE_RESOURCE_ASSIGNMENT_V4_IP_LOCAL_ADDRESS].value.uint32;
+
+    ctx.process_id = incoming_metadata_values->processId;
+    memcpy(&ctx.socket_address, &addr, sizeof(addr));
+    ctx.operation = BIND_OPERATION_BIND;
+    ctx.protocol = incoming_fixed_values->incomingValue[FWPS_FIELD_ALE_RESOURCE_ASSIGNMENT_V4_IP_PROTOCOL].value.uint8;
+
+    ctx.app_id_start =
+        incoming_fixed_values->incomingValue[FWPS_FIELD_ALE_RESOURCE_ASSIGNMENT_V4_ALE_APP_ID].value.byteBlob->data;
+    ctx.app_id_end =
+        ctx.app_id_start +
+        incoming_fixed_values->incomingValue[FWPS_FIELD_ALE_RESOURCE_ASSIGNMENT_V4_ALE_APP_ID].value.byteBlob->size;
+
+    _net_ebpf_ext_resource_truncate_appid(&ctx);
+    if (ebpf_ext_attach_invoke_hook(_ebpf_bind_hook_provider_registration, &ctx, &result) == EBPF_SUCCESS) {
+        switch (result) {
+        case BIND_PERMIT:
+        case BIND_REDIRECT:
+            classify_output->actionType = FWP_ACTION_PERMIT;
+            break;
+        case BIND_DENY:
+            classify_output->actionType = FWP_ACTION_BLOCK;
+        }
+    }
+
+Exit:
+    ebpf_ext_attach_leave_rundown(_ebpf_bind_hook_provider_registration);
+    return;
+}
+
+void
+net_ebpf_ext_resource_release_classify(
+    _In_ const FWPS_INCOMING_VALUES* incoming_fixed_values,
+    _In_ const FWPS_INCOMING_METADATA_VALUES* incoming_metadata_values,
+    _Inout_opt_ void* layer_data,
+    _In_opt_ const void* classify_context,
+    _In_ const FWPS_FILTER* filter,
+    uint64_t flow_context,
+    _Inout_ FWPS_CLASSIFY_OUT* classify_output)
+{
+    SOCKADDR_IN addr = {AF_INET};
```

Comment:
IPv6?

Reply:
https://github.com/microsoft/ebpf-for-windows/issues/526